step-controller-plugins/step-controller-measurement/src/main/java/step/plugins/measurements/MeasurementPlugin.java

```diff
@@ -97,18 +101,48 @@ public void run() {
 	@Override
 	public void beforeReportNodeExecution(ExecutionContext context, ReportNode node) {
 		if (node instanceof ThreadReportNode) {
-			ThreadReportNode tNode = (ThreadReportNode) node;
-			GaugeCollector gaugeCollector = gaugeCollectorRegistry.getGaugeCollector(MeasurementControllerPlugin.ThreadgroupGaugeName);
-			String schedulerTaskId = (String) context.get(SCHEDULER_TASK_ID);
-			schedulerTaskId = (schedulerTaskId!=null) ? schedulerTaskId : "";
-			String planId = context.getPlan().getId().toString();
-			String[] labels = {tNode.getExecutionID(),tNode.getThreadGroupName(),planId,schedulerTaskId};
-			gaugeCollector.getGauge().labels(labels).inc();
-			labelsByExec.get(context.getExecutionId()).add(labels);
-			List<Measurement> measurements = gaugeCollector.collectAsMeasurements();
-			for (MeasurementHandler measurementHandler : MeasurementPlugin.measurementHandlers) {
-				measurementHandler.processGauges(measurements);
-			}
+			processThreadReportNode(context, (ThreadReportNode) node, true);
+		}
+	}
+
+	@Synchronized
+	private GaugeCollector getOrInitThreadGauge(ExecutionContext context) {
+		GaugeCollector gaugeCollector = gaugeCollectorRegistry.getGaugeCollector(ThreadgroupGaugeName);
```

Comment:
This code is specific to the ThreadGroup artefact. Ideally we should do this in a "pluggable" way and move this closer to the code of the ThreadGroup.

Reply:
indeed, didn't want to refactor this as part of this change.